Add token to github API call in CI

pfebrer opened the pull request:
Every once in a while the CI fails because we are sending an authenticated request to the github API, which has rate limits.

In this PR I use our GH_READ_TOKEN for authenticating so that this doesn't happen anymore

Luthaf commented on .github/workflows/docs.yml:
```diff
@@ -27,13 +27,14 @@ jobs:
       - id: gatherInfo
         # Collect the examples that need to be run, and the id of the latest
         # workflow run on the main branch.
+        permissions: read
```
Is this required? I think `read` is the default anyway

pfebrer replied:
Hmm no read is not the default, because GITHUB_TOKEN until now has been used to deploy to gh-pages. In PRs that come from forks however GITHUB_TOKEN can only read, so the protection against attacks from PR submitters is already there by default, but not against github actions having some malicious code.

Luthaf replied:
ok, fair enough! Although the only scenario where I would see someone trying to use malicious code would be a PR from a fork anyway

pfebrer replied:
Well, the code in a Github Action can do whatever it wants really, so if you give them the token they could in principle do whatever they want with your repo :sweat_smile:

Luthaf commented on .github/workflows/docs.yml:
```diff
@@ -99,10 +100,11 @@ jobs:
       - name: Download latest main branch built examples
         uses: actions/download-artifact@v4
         if: ${{ github.event_name == 'pull_request' }}
+        permissions: read
```
same

Luthaf reviewed (commented):
I'll generate a token now

pfebrer committed:
Improve handling of tokens in CI.

This commit fixes two problems:
1. We were hitting API rate limits when sending
unauthenticated requests.
2. The builds for the unmodified examples during a PR
were not possible to download if the PR was from a fork.
